torch/_inductor/fx_passes/split_cat.py

```diff
@@ -48,20 +57,125 @@ def apply(self, match, graph, node):
             # TODO dynamic_shapes with assume_static_by_default=False fails while AOT Autograd tracing.
             return
+        if split_dim < 0:  # Normalize split dim
+            split_dim += split_input.meta["example_value"].dim()
         with graph.inserting_after(split_node):
             new_split_node = graph.call_function(
-                torch.split, args=(split_input, split_sections, split_dim)
+                torch.split,
+                args=(split_input, split_sections),
+                kwargs={"dim": split_dim},
             )
         split_node.replace_all_uses_with(new_split_node)
         new_split_node.meta.update(split_node.meta)
         graph.erase_node(split_node)
         counters["inductor"]["split_cat_norm"] += 1
 
 
+class TorchSplit(CallFunction):
+    """
+    Matches a call to torch.split if it is in a normalized form. Ensures that all users of
+    splits are unique getitems.
+    """
+
+    def __init__(self, arg):
+        # using KeywordArg("dim") for `dim` checks they all match
+        super().__init__(
+            torch.split, arg, Arg(), _users=MULTIPLE, dim=KeywordArg("dim")
+        )
+
+    def _match(self, node: torch.fx.Node, ctx: MatchContext):
+        m = super()._match(node, ctx)
+        if not m:
+            return m
+        split_sections = get_arg_value(node, 1, "split_size_or_sections")
+        if not isinstance(split_sections, (list, tuple)):
+            return FailedMatch("split not normalized")
+        # check users are all unique getitems
+        seen_idxs = set()
+        for user in node.users:
+            if not match_node(user, {operator.getitem}):
+                # This should ideally never happen. Split user should always be a getitem
+                return FailedMatch(f"user of split not a getitem: {user}")
+            if user.args[1] in seen_idxs:
+                return FailedMatch(f"duplicate getitem {user.args[1]}")
+            if user.args[-1] < 0:
+                # This shouldn't ideally happen as dynamo normalizes indexes to positive
+                return FailedMatch("negative index")
+            seen_idxs.add(user.args[1])
+        return m
+
+
+class MergeConsecutiveSplits(PatternEntry):
+    def apply(self, match: Match, graph: torch.fx.Graph, node: torch.fx.Node):
+        first_split = node.args[0].args[0]
+        first_split_sections = get_arg_value(first_split, 1, "split_size_or_sections")
+        next_split_sections = get_arg_value(node, 1, "split_size_or_sections")
+        next_split_index = node.args[0].args[1]
+
+        new_split_sections = list(first_split_sections)
+        new_split_sections[
+            next_split_index : next_split_index + 1
+        ] = next_split_sections
+
+        first_split_input = first_split.args[0]
+        first_split_dim = get_arg_value(first_split, 2, "dim")
+
+        to_remove = []
+
+        with graph.inserting_after(first_split_input):
```

Comment:
Shouldn't this be before the first split node to preserve graph order?

Reply:
Being after the split's input should also preserve graph order right? Since split can have only one tensor input?